tools/api_proto_breaking_change_detector/detector.py

```diff
@@ -0,0 +1,147 @@
+from abc import ABC, abstractmethod
+import tempfile
+from rules_python.python.runfiles import runfiles
+from tools.run_command import run_command
+from shutil import copyfile
+import re
+from pathlib import Path
+import os
+"""
+This tool is used to detect "breaking changes" in protobuf files, to ensure proper backwards-compatibility in protobuf API updates.
+The tool can check for breaking changes of a single API by taking 2 .proto file paths as input (before and after) and outputting a bool `is_breaking`.
+
+The breaking change detector creates a temporary directory, copies in each file to compute a protobuf "state",
+computes a diff of the "before" and "after" states, and runs the diff against a set of rules to determine if there was a breaking change.
+
+The tool is currently implemented with buf (https://buf.build/)
+"""
+
+
+# generic breaking change detector for protos, extended by a wrapper class for a breaking change detector
+class ProtoBreakingChangeDetector(ABC):
+
+    @abstractmethod
+    def __init__(self, path_to_before, path_to_after, additional_args=None):
+        pass
+
+    @abstractmethod
+    def run_detector(self):
+        pass
+
+    @abstractmethod
+    def is_breaking(self):
+        pass
+
+    @abstractmethod
+    def lock_file_changed(self):
+        pass
+
+
+class ChangeDetectorError(Exception):
+    pass
+
+
+class ChangeDetectorInitializeError(ChangeDetectorError):
+    pass
+
+
+BUF_LOCK_FILE = "tmp.json"
+
+
+class BufWrapper(ProtoBreakingChangeDetector):
+
+    def __init__(self, path_to_before, path_to_after, additional_args=None):
+        if not Path(path_to_before).is_file():
+            raise ValueError(f"path_to_before {path_to_before} does not exist")
+
+        if not Path(path_to_after).is_file():
+            raise ValueError(f"path_to_after {path_to_after} does not exist")
+
+        self.path_to_before = path_to_before
+        self.path_to_after = path_to_after
+        self.additional_args = additional_args
+
+    def run_detector(self):
+        # 1) pull buf BSR dependencies with buf mod update (? still need to figure this out. commented out for now)
```

Comment:
Does this need to be done before every invocation of the detector, or just once before all detectors are run?

Reply:
Once per invocation (although we can introduce separation of environment setup and running buf once I add in the logic that pulls the BSR dependencies)